src/rules/noUnsafeFinallyRule.ts

```diff
@@ -40,171 +39,112 @@ export class Rule extends Lint.Rules.AbstractRule {
-
-    /**
-     * A value indication whether the current scope is a `return` and `throw` boundary.
-     */
-    isReturnsThrowsBoundary: boolean;
-
-    /**
-     * A collection of `break` or `continue` labels in this scope.
-     */
-    labels: string[];
-}
-
-/**
- * Represents a block walker that identifies finally blocks and walks
- * only the blocks that do not change scope for return statements.
- */
-class NoReturnInFinallyScopeAwareWalker extends Lint.ScopeAwareRuleWalker<IFinallyScope> {
-
-    protected visitBreakStatement(node: ts.BreakOrContinueStatement) {
-        if (!this.isControlFlowWithinFinallyBlock(isBreakBoundary, node)) {
-            super.visitBreakStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_BREAK));
-    }
-
-    protected visitContinueStatement(node: ts.BreakOrContinueStatement) {
-        if (!this.isControlFlowWithinFinallyBlock(isContinueBoundary, node)) {
-            super.visitContinueStatement(node);
-            return;
-        }
-
-        this.addFailureAtNode(node, Rule.FAILURE_STRING_FACTORY(Rule.FAILURE_TYPE_CONTINUE));
-    }
-
-    protected visitLabeledStatement(node: ts.LabeledStatement) {
-        this.getCurrentScope().labels.push(node.label.text);
-
-        super.visitLabeledStatement(node);
-    }
-
-    protected visitReturnStatement(node: ts.ReturnStatement): void {
-        if (!this.isControlFlowWithinFinallyBlock(isReturnsOrThrowsBoundary)) {
-            super.visitReturnStatement(node);
-            return;
+type JumpStatement = ts.BreakStatement | ts.ContinueStatement | ts.ThrowStatement | ts.ReturnStatement;
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    let inFinally = false;
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        switch (node.kind) {
+            case ts.SyntaxKind.TryStatement:
+                const { tryBlock, catchClause, finallyBlock } = node as ts.TryStatement;
+                ts.forEachChild(tryBlock, cb);
+                if (catchClause !== undefined) {
+                    ts.forEachChild(catchClause, cb);
+                }
+                if (finallyBlock !== undefined) {
+                    const old = inFinally;
+                    inFinally = true;
+                    cb(finallyBlock);
+                    inFinally = old;
+                }
+                break;
+
+            case ts.SyntaxKind.BreakStatement:
+            case ts.SyntaxKind.ContinueStatement:
+            case ts.SyntaxKind.ThrowStatement:
+            case ts.SyntaxKind.ReturnStatement:
+                if (inFinally && !jumpIsLocalToFinallyBlock(node as JumpStatement)) {
+                    ctx.addFailureAtNode(node, Rule.FAILURE_STRING(printJumpKind(node as JumpStatement)));
+                }
+                ts.forEachChild(node, cb);
+                break;
+
+            default:
+                ts.forEachChild(node, cb);
```

Comment:
I don't think that's necessary. `jumpIsLocalToFinallyBlock` checks for function scopes.